src/envoy/alts/tsi_transport_socket.cc

```diff
@@ -0,0 +1,207 @@
+    : handshaker_factory_(handshaker_factory), raw_buffer_callbacks_(*this) {
+  raw_buffer_socket_.setTransportSocketCallbacks(raw_buffer_callbacks_);
+}
+
+TsiSocket::~TsiSocket() { ASSERT(!handshaker_); }
+
+void TsiSocket::setTransportSocketCallbacks(
+    Envoy::Network::TransportSocketCallbacks &callbacks) {
+  callbacks_ = &callbacks;
+
+  handshaker_ = handshaker_factory_(callbacks.connection().dispatcher());
+  handshaker_->setHandshakerCallbacks(*this);
+}
+
+std::string TsiSocket::protocol() const { return ""; }
+
+Network::PostIoAction TsiSocket::doHandshake() {
+  ASSERT(!handshake_complete_);
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake", callbacks_->connection());
+
+  if (!handshaker_next_calling_) {
+    doHandshakeNext();
+  }
+  return Network::PostIoAction::KeepOpen;
+}
+
+void TsiSocket::doHandshakeNext() {
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake next: received: {}",
+                 callbacks_->connection(), raw_read_buffer_.length());
+  handshaker_next_calling_ = true;
+  Buffer::OwnedImpl handshaker_buffer;
+  handshaker_buffer.move(raw_read_buffer_);
+  handshaker_->next(handshaker_buffer);
+}
+
+Network::PostIoAction TsiSocket::doHandshakeNextDone(
+    NextResultPtr &&next_result) {
+  ASSERT(next_result);
+
+  ENVOY_CONN_LOG(debug, "TSI: doHandshake next done: status: {} to_send: {}",
+                 callbacks_->connection(), next_result->status_,
+                 next_result->to_send_->length());
+
+  tsi_result status = next_result->status_;
+  tsi_handshaker_result *handshaker_result = next_result->result_.get();
+
+  if (status != TSI_INCOMPLETE_DATA && status != TSI_OK) {
+    ENVOY_CONN_LOG(debug, "TSI: Handshake failed: status: {}",
+                   callbacks_->connection(), status);
+    return Network::PostIoAction::Close;
+  }
+
+  if (next_result->to_send_->length() > 0) {
+    raw_write_buffer_.move(*next_result->to_send_);
+  }
+
+  if (status == TSI_OK && handshaker_result != nullptr) {
+    tsi_peer peer;
+    tsi_handshaker_result_extract_peer(handshaker_result, &peer);
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: peer properties: {}",
+                   callbacks_->connection(), peer.property_count);
+    for (size_t i = 0; i < peer.property_count; ++i) {
+      ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(),
+                     peer.properties[i].name,
+                     std::string(peer.properties[i].value.data,
+                                 peer.properties[i].value.length));
+    }
+    tsi_peer_destruct(&peer);
+
+    const unsigned char *unused_bytes;
+    size_t unused_byte_size;
+
+    status = tsi_handshaker_result_get_unused_bytes(
+        handshaker_result, &unused_bytes, &unused_byte_size);
+    ASSERT(status == TSI_OK);
+    if (unused_byte_size > 0) {
+      raw_read_buffer_.add(unused_bytes, unused_byte_size);
+    }
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: unused_bytes: {}",
```

Comment:
Would be nice to document why/what are these unused byte buffer.

Reply:
It is clear from tsi_handshaker doc, the bytes not used for handshake.